contracts/0.8.25/vaults/StakingVault.sol

```diff
@@ -404,6 +335,175 @@ contract StakingVault is IStakingVault, OwnableUpgradeable {
+     */
+    function depositToBeaconChain(Deposit[] calldata _deposits) external {
+        if (_deposits.length == 0) revert ZeroArgument("_deposits");
+
+        ERC7201Storage storage $ = _getStorage();
+
+        if (msg.sender != $.nodeOperator) revert NotAuthorized("depositToBeaconChain", msg.sender);
+        if ($.beaconChainDepositsPaused) revert BeaconChainDepositsArePaused();
+        if (valuation() < $.locked) revert Unbalanced();
+
+        uint256 totalAmount = 0;
+        uint256 numberOfDeposits = _deposits.length;
+        for (uint256 i = 0; i < numberOfDeposits; i++) {
+            IStakingVault.Deposit calldata deposit = _deposits[i];
+            BEACON_CHAIN_DEPOSIT_CONTRACT.deposit{value: deposit.amount}(
+                deposit.pubkey,
+                bytes.concat(withdrawalCredentials()),
+                deposit.signature,
+                deposit.depositDataRoot
+            );
+            totalAmount += deposit.amount;
+        }
+
+        emit DepositedToBeaconChain(msg.sender, numberOfDeposits, totalAmount);
+    }
+
+    /**
+     * @notice Calculates the total withdrawal fee required for given number of validator keys
+     * @param _numberOfKeys Number of validators' public keys
+     * @return Total fee amount
+     */
+    function calculateValidatorWithdrawalFee(uint256 _numberOfKeys) external view returns (uint256) {
+        if (_numberOfKeys == 0) revert ZeroArgument("_numberOfKeys");
+
+        return _numberOfKeys * TriggerableWithdrawals.getWithdrawalRequestFee();
+    }
+
+    /**
+     * @notice Requests validator exit from the beacon chain by emitting an `ValidatorExitRequested` event
+     * @param _pubkeys Concatenated validators' public keys
+     * @dev    Signals the node operator to eject the specified validators from the beacon chain
+     */
+    function requestValidatorExit(bytes calldata _pubkeys) external onlyOwner {
+        if (_pubkeys.length == 0) revert ZeroArgument("_pubkeys");
+
+        emit ValidatorExitRequested(msg.sender, _pubkeys);
+    }
+
+    /**
+     * @notice Initiates a full validator withdrawal from the beacon chain following EIP-7002
+     * @param _pubkeys Concatenated validators public keys
+     * @dev    Keys are expected to be 48 bytes long tightly packed without paddings
+     *         Only allowed to be called by the owner or the node operator
+     * @dev    The caller must provide sufficient fee via msg.value to cover the withdrawal request costs
+     */
+    function initiateFullValidatorWithdrawal(bytes calldata _pubkeys) external payable {
+        if (_pubkeys.length == 0) revert ZeroArgument("_pubkeys");
+        _onlyOwnerOrNodeOperator();
+
+        (uint256 feePerRequest, uint256 totalFee) = _getAndValidateWithdrawalFees(_pubkeys);
+        TriggerableWithdrawals.addFullWithdrawalRequests(_pubkeys, feePerRequest);
+
+        emit FullValidatorWithdrawalInitiated(msg.sender, _pubkeys);
```

Comment:
let's not split this event and add refund

Reply:
No need to split this event.